### PR TITLE
Add client instances

### PR DESCRIPTION
**Pull request by tredman**
For beelines, we're likely going to want to initialize libhoney in the beeline in a way that could conflict with how the consumer is already using libhoney. There may also be places in the beeline where we need to init libhoney, but can't know if it's already initialized in other places. It's safer if we can have a client instance that's isolated from what's going on in the global state.

### CODE REVIEW

File: libhoney/__init__.py [hunk 1]
```diff
@@ -356,11 +362,20 @@ def send_presampled(self):
                 "No api_host for Honeycomb. Can't send to the Great Unknown.")
         if self.writekey == "":
             raise SendError(
-                "No write_key specified. Can't send event.")
+                "No writekey specified. Can't send event.")
```
Comment (asdvalenzuela) [inline libhoney/__init__.py]: write key

**Comment by tredman** [comment]
Refactored this to replace the mix of global state variables with a single global client. I had to break out classes into their own modules due a circular import issue I ran into. I did this with an eye for not breaking the existing API but I would like more :eyes: on that.

File: libhoney/errors.py [hunk 2]
```diff
@@ -0,0 +1,7 @@
+class SendError(Exception):
+    def __init__(self, *args, **kwargs):
+        Exception.__init__(self, *args, **kwargs)
```
[the remark this answers is not in the document]
Comment (tredman) [reply libhoney/errors.py]: Yeah the usual way to create custom exceptions is to just derive from `Exception` and pass.

**Comment by tredman** [comment]
Hmm - what if we just stopped throwing `NotInitializedError` and returned None/made the op a no-op? Could still break things but in less terrible ways than throwing a new exception. I need to see how other libraries log these kinds of problems (i.e. "you're attempting to use our lib, but doing it wrong, so we're logging here instead of failing!").

Agree on the recommendation of clients over global scope.

**Comment by tredman** [comment]
Ok. I got rid of `NotInitializedError` and replaced it with a one-time warning. If global state is not initialized (or a client is not used), everything should behave as before, up until an attempt to call `Event.send` is made. At that time, we will raise a `SendError` like we've been doing.

**Comment by samstokes** [comment]
I think replacing `NotInitializedError` with `_warn_uninitialized` then falling back to no-op is a good change. If I read this correctly though, this retains the runtime `SendError` on sending an event? IMHO that's the case _most_ deserving of being converted into a no-op (since it's the thing that breaks your app in production if you forget to initialize libhoney).

libhoney-rb previously made a large customer unhappy because it spewed thousands of errors per second to their production logs in a similar circumstance (although not quite the same, since that was due to misconfiguration rather than omitted initialization).

Again, this shouldn't block the merging of this PR which is a clear improvement; but I think it would be better if `Event.send()` was *also* a no-op if the client was uninitialized, probably with the same `_warn_uninitialized` call. WDYT?

**Comment by tredman** [comment]
Yeah, I like that. Do we still want to throw SendErrors in other cases?

```
if self._fields.is_empty():
            raise SendError(
                "No metrics added to event. Won't send empty event.")
        if self.api_host == "":
            raise SendError(
                "No api_host for Honeycomb. Can't send to the Great Unknown.")
        if self.writekey == "":
            raise SendError(
                "No writekey specified. Can't send event.")
        if self.dataset == "":
            raise SendError(
                "No dataset for Honeycomb. Can't send event without knowing which dataset it belongs to.")
```

**Comment by samstokes** [comment]
> Do we still want to throw SendErrors in other cases? Seems like all of these are undesirable from a customer perspective.

I think we do not want to throw SendErrors in those cases either, and I [recently changed libhoney-rb](https://github.com/honeycombio/libhoney-rb/pull/25) to fall back to a no-op in those cases. However, the user story there is "remembered to initialize libhoney but didn't fill in some of its config, then tried to send events", which is slightly less clear-cut than "forgot to initialize libhoney at all, then tried to send events". So I suggest leaving that change out of the scope of this PR.

There's a broader conversation to be had about the SDK, how we handle misconfiguration, and how we support certain scenarios where you omit the write key and/or dataset from the `.Init` call (or `Client` instantiation) and specify them at the builder level instead. My libhoney-rb change linked above jumped the gun and presupposed a conclusion to that conversation, but [Ben pointed out](https://houndsh.slack.com/archives/C4M02EG83/p1528221051000272) that makes it inconsistent with the other SDKs and with our [SDK spec](https://docs.honeycomb.io/reference/sdk-spec/). I have a todo item to start an email thread to discuss this.

**Comment by samstokes** [comment]
> So I suggest leaving that change out of the scope of this PR.

To clarify, "that change" = specifically the change to make `send` a no-op in case of missing configuration. Making it a no-op in the uninitialized case is definitely a good idea.

**Pull request merged**